engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeAnyNode.java

```diff
@@ -79,14 +80,32 @@ long hashCodeForFloat(float f) {
   }
 
   @Specialization
+  @TruffleBoundary
   long hashCodeForDouble(double d) {
-    return d % 1.0 == 0.0 ? (int) d : Double.hashCode(d);
+    if (d % 1.0 != 0.0) {
+      return Double.hashCode(d);
+    } else {
+      long longValue = Double.valueOf(d).longValue();
+      if (BigIntegerOps.fitsInLong(d)) {
+        return Long.hashCode(longValue);
```

Comment:
1. `longValue` is not needed in the other branch, so I don't see why compute it before the branch (unless there is some reason, then pls ignore)
2. I think it may be cleaner to delegate to `hashCodeForLong` - then if the implementation ever changed, we will get consistency for free. Can we make this method `final`? Then that should not incur any additional cost.

```suggestion
      if (BigIntegerOps.fitsInLong(d)) {
        long longValue = Double.valueOf(d).longValue();
        return hashCodeForLong(longValue);
```

Reply:
> longValue is not needed in the other branch, so I don't see why compute it before the branch (unless there is some reason, then pls ignore)

No, I have just overlooked this.

> I think it may be cleaner to delegate to hashCodeForLong - then if the implementation ever changed, we will get consistency for free. Can we make this method final? Then that should not incur any additional cost.

Delegating to `hashCodeForLong` definitely sounds better. I have not seen any truffle dev making specialization methods `final`, I doubt that would get us any better performance. I assume that Graal speculates on these methods being effectively final, but I may be wrong.